### ISSUE TITLE
High cpu usage with Tab Stash

### ISSUE THREAD

**Issue by keithx2** [josh-berry/tab-stash#240]
This happens on my main profiles, but I also created a new profile and still got high cpu usage (usually 50%+).

Firefox settings with new profile:

1)  Open previous windows and tabs option is selected (about:preferences).
2)  Sync is off.

Tab Stash settings with new profile:

1) Show advanced settings
2) Stash the active tab and don't show the stash
3) Don't show the stash
4) When stashing a single tab: ... 600 minutes
5) Hide the tab and immediately unload it

I have the bookmark sidebar open, have the stash shown in a tab, and I have the Unstashed Tabs group collapsed.  Then I stash a few tabs and restart Firefox.  The cpu usage is then really high.  If it's not high, I'll stash a few more tabs, and restart Firefox again.  It eventually happens for me.  If I disable Tab Stash, the cpu usage will immediately drop.

OS - Windows 10
Firefox - 99 beta and 100 alpha
Tab Stash - 2.10.3

Console error:

Security Error: Content at moz-extension://.../stash-list.html may not load or link to chrome://mozapps/skin/extensions/extension.svg.

<hr>
<em>Readers: If you are also experiencing this bug, please vote for it by giving the ORIGINAL POST a thumbs-up using the :smiley: button.  You are welcome to leave comments and discuss the issue if you have more details to add, but "Me too!" comments are not counted by the voting system.</em>

**Comment by keithx2** [comment]
Thanks for responding.

I'm using the Windows Task Manager to monitor cpu usage.
I tested a little more and found the following steps to give me high cpu usage:

1)  Create a new Firefox profile
2)  Install Tab Stash
3)  Select 'Hide the tabs for this group' for the Unstashed Tabs in the sidebar
4)  Restart Firefox

After that point, no matter what I do, the cpu usage stays high while Tab Stash is enabled.

**Comment by KerfuffleV2** [comment]
> Thanks for responding.

Sure.

> I'm using the Windows Task Manager to monitor cpu usage.

You might also want to look at `about:performance`. It can show more detail about which specific components in Firefox are using CPU.

> After that point, no matter what I do, the cpu usage stays high while Tab Stash is enabled.

I'm too lazy to create a new profile but I have 12 groups and about 200 tabs and expanding/collapsing groups definitely doesn't cause permanent CPU usage. If I spam click the master toggle near the search which shows/hides 200 tabs and bookmarks I can get it to use a noticeable amount of CPU but it drops off immediately once I stop. I'm using the latest Nightly, for reference.

I don't use Firefox's sync feature. I know it can do stuff like sync open tabs, which seems like something that could be expensive. If you use sync, could you try disabling it temporarily and see if that helps? (Probably restart the browser after disabling sync.)

**Comment by KerfuffleV2** [comment]
@keithx2 @josh-berry I found and fixed the issue. Seems like it's a very long standing bug and probably what previous reports of TS CPU usage came from also. Seems like that issue was introduced about 2 years ago.

Pretty tricky one to track down, but I knew it had to be something with bookmark metadata and the tabs group! I'm surprised that FF's performance monitoring stuff didn't show it was TS using CPU.

**Comment by keithx2** [comment]
Nice :)  Thank you for the quick work!

**Comment by josh-berry** [comment]
Sorry @keithx2 for the trouble and thanks @KerfuffleV2 for looking into it and fixing it! I've merged the fix and it will be in the next release.

@KerfuffleV2 I would guess FF didn't show the CPU usage because most of the time would have actually been spent querying (and re-querying, and re-querying, and re-querying...) the KVS. So most of the time would have gone into the Firefox runtime (message-passing) and IndexedDB.

------------------

In the meantime, anyone hitting this issue can work around it by creating a stash, collapsing and expanding it. So long as there's at least one stash that has some data saved about whether it should be collapsed/expanded (i.e. the "hide/show" arrow has been clicked at least once), you shouldn't hit this bug.

**Comment by mt-empty** [comment]
> In the meantime, anyone hitting this issue can work around it by creating a stash, collapsing and expanding it. So long as there's at least one stash that has some data saved about whether it should be collapsed/expanded (i.e. the "hide/show" arrow has been clicked at least once), you shouldn't hit this bug.

Could you elaborate on this? ELI5
Thank you

**Comment by KerfuffleV2** [comment]
@mt-empty 
First, if you're not seeing Firefox use excessive CPU that stops immediately if you disable the Tab Stash addon then there's no need to worry about this at all and no action is required. This is also a fairly uncommon issue because you have to have TS installed but not actually use it. Most people that install the addon would actually stash something. Anyway...

There was a rare issue where TS could use excessive CPU if you started the browser with no stash groups at all. The way to work around that in the current version is to ensure there is always at least one tab stashed (so at least one group exists) and also collapse/expand that group before exiting Firefox so the setting for group state exists. 

This will be fixed in the next release version (2.11, I think).

**Comment by mt-empty** [comment]
@KerfuffleV2 
I'm on Ubuntu and I'm getting this issue even though I have plenty of stashed group.
Simply enabling the TS - by going to about:addons - without doing anything else, greatly increases CPU usage.

I tried what you said but it didn't help. I currently have it disabled, I guess I will wait for the next release and check if the issue persists.

**Comment by herbdool** [comment]
I had same problem on Ubuntu. Uninstalling and reinstalling seems to have helped. Previously CPU usage as around 13%, dropped to < 0.1%. Memory from around 300 MB to 68MB (includes a couple other extensions).

**Comment by icemtel** [comment]
I had the same problem as discussed above: high CPU usage; which appears immediately after the Tab Stash is turned on.
Took me a long time to figure out, as nothing was shown in `about:performance`. 

Re-installing the extension solved the issue (at least for now).
